### PR TITLE
Feature/dhcp processors

### PR DESCRIPTION
**Pull request by dwlfrth**
# Description
Move common flows / process of dhcp processors in base class

# Impacts
* fingerbank processing
* dhcp filtering
* pretty much in ip task of dhcp flow

# Delete branch after merge
YES

# UPGRADE file entries
* dhcp_filters.conf -> 'computer_name' attribute name changed to 'computername'
* dhcp_filters.conf -> 'DhcpFingerbank' scope changed to 'Fingerbank'

### CODE REVIEW

**Comment by dwlfrth** [comment]
Review me

File: lib/pf/api.pm [hunk 1]
```diff
@@ -216,19 +216,45 @@ sub update_iplog : Public :AllowedAsAction(mac, $mac, ip, $ip) {
...
-            "oldip ($postdata{'oldip'}) and newip ($postdata{'ip'}) are different for $postdata{'mac'} - closing iplog entry"
+            "oldip ($postdata{'oldip'}) and newip ($postdata{'ip'}) are different for $postdata{'mac'} - closing ip4log entry"
         );
         pf::ip4log::close($postdata{'oldip'});
     }
 
     return (pf::ip4log::open($postdata{'ip'}, $postdata{'mac'}, $postdata{'lease_length'}));
 }
 
+sub update_ip6log : Public :AllowedAsAction(mac, $mac, ip, $ip) {
```
Comment (julsemaan) [inline lib/pf/api.pm]: perldoc

File: lib/pf/api.pm [hunk 2]
```diff
@@ -203,7 +203,7 @@ sub radius_switch_access : Public {
     return $return;
 }
 
-sub update_iplog : Public :AllowedAsAction(mac, $mac, ip, $ip) {
+sub update_ip4log : Public :AllowedAsAction(mac, $mac, ip, $ip) {
```
Comment (julsemaan) [inline lib/pf/api.pm]: perldoc

File: lib/pf/dhcp/processor.pm [hunk 3]
```diff
@@ -19,40 +19,178 @@ use warnings;
...
+    # Firewall SSO
+    if ( $iptasks_arguments{'oldip'} && $iptasks_arguments{'oldip'} ne $iptasks_arguments{'ip'} ) {
+        $self->apiClient->notify( 'firewallsso', (method => 'Stop', mac => $iptasks_arguments{'mac'}, ip => $iptasks_arguments{'oldip'}, timeout => undef) );
+        $self->apiClient->notify( 'firewallsso', (method => 'Start', mac => $iptasks_arguments{'mac'}, ip => $iptasks_arguments{'ip'}, timeout => $iptasks_arguments{'lease_length'} || $DEFAULT_LEASE_LENGTH) );
+    }
+    $self->apiClient->notify( 'firewallsso', (method => 'Update', mac => $iptasks_arguments{'mac'}, ip => $iptasks_arguments{'ip'}, timeout => $iptasks_arguments{'lease_length'} || $DEFAULT_LEASE_LENGTH) );
+
+    # Inline enforcement
+    # 2017.03.20 - dwuelfrath@inverse.ca - There is currently no ipv6 support for inline enforcement. Remove the condition once "resolved"
+    unless ( $iptasks_arguments{'ipversion'} eq "ipv6" ) {
```
Comment (julsemaan) [inline lib/pf/dhcp/processor.pm]: ipv6 as a constant

File: lib/pf/dhcp/processor.pm [hunk 4]
```diff
@@ -19,40 +19,178 @@ use warnings;
...
+        if ( $iptasks_arguments{'oldip'} && $iptasks_arguments{'oldip'} ne $iptasks_arguments{'ip'} ) {
+            my $node_view = node_view($iptasks_arguments{'mac'});
+            my $last_connection_type = $node_view->{'last_connection_type'};
+            $self->apiClient->notify('ipset_node_update', $iptasks_arguments{'oldip'}, $iptasks_arguments{'ip'}, $iptasks_arguments{'mac'}) if (defined $last_connection_type && $last_connection_type eq $connection_type_to_str{$INLINE});
+        }
+    }
+
+    # Conformity scan
+    # 2017.03.20 - dwuelfrath@inverse.ca - There is currently no ipv6 support for conformity scan. Remove the condition once "resolved"
+    unless ( $iptasks_arguments{'ipversion'} eq "ipv6" ) {
```
Comment (julsemaan) [inline lib/pf/dhcp/processor.pm]: again

File: lib/pf/dhcp/processor.pm [hunk 5]
```diff
@@ -19,40 +19,178 @@ use warnings;
...
+
+    # Parking violation
+    $self->checkForParking($iptasks_arguments{'mac'}, $iptasks_arguments{'ip'});
+    if ( $iptasks_arguments{'oldmac'} && $iptasks_arguments{'oldmac'} ne $iptasks_arguments{'mac'} ) {
+        # Remove the actions that were for the previous MAC address
+        pf::parking::remove_parking_actions($iptasks_arguments{'oldmac'}, $iptasks_arguments{'ip'});
+    }
+
+    # IPlog
+    if ( $iptasks_arguments{'ipversion'} eq "ipv4" ) {
```
Comment (julsemaan) [inline lib/pf/dhcp/processor.pm]: ipv4 as constant

File: lib/pf/dhcp/processor.pm [hunk 6]
```diff
@@ -19,40 +19,178 @@ use warnings;
...
+        $self->apiClient->notify('update_ip4log', %iptasks_arguments);
+    } elsif ( $iptasks_arguments{'ipversion'} eq "ipv6" ) {
```
Comment (julsemaan) [inline lib/pf/dhcp/processor.pm]: constant

File: lib/pf/dhcp/processor.pm [hunk 7]
```diff
@@ -19,40 +19,178 @@ use warnings;
...
+    }
+
+    # Conformity scan
+    # 2017.03.20 - dwuelfrath@inverse.ca - There is currently no ipv6 support for conformity scan. Remove the condition once "resolved"
+    unless ( $iptasks_arguments{'ipversion'} eq "ipv6" ) {
+        $self->apiClient->notify('trigger_scan', %iptasks_arguments );
+    }
+
+    # Parking violation
+    $self->checkForParking($iptasks_arguments{'mac'}, $iptasks_arguments{'ip'});
```
Comment (julsemaan) [inline lib/pf/dhcp/processor.pm:151]: Should probably only do this in ipv4 as the parking won't be able to handle ipv6 in OMAPI, ipset, ...

File: lib/pf/dhcp/processor_v4.pm [hunk 8]
```diff
@@ -667,98 +644,37 @@ sub parse_dhcp_option82 {
...
-    $self->{api_client}->notify('update_iplog', %data);
-}
-
-=head2 get_local_dhcp_servers_by_ip
-
-Return a list of all dhcp servers IP that could be running locally.
 
-Caches results on first run then returns from cache.
+    # Add IP version to arguments
+    $iptasks_arguments->{'ipversion'} = "ipv4";
```
Comment (julsemaan) [inline lib/pf/dhcp/processor_v4.pm]: I heard about a great city, its called Constantinople.
Maybe we should bring the Inverse team there for the summer vacations

File: lib/pf/dhcp/processor_v6.pm [hunk 9]
```diff
@@ -201,6 +226,48 @@ sub _process_options {
...
+    my $mac = $iptasks_arguments->{'mac'};
+
+    # Sanitize input
+    unless ( pf::util::valid_mac($mac) || pf::util::IP::is_ipv6($ip) ) {
+        $logger->error("invalid MAC or IP: $mac $ip");
+        return;
+    }
+
+    # Add IP version to arguments
+    $iptasks_arguments->{'ipversion'} = "ipv6";
```
Comment (julsemaan) [inline lib/pf/dhcp/processor_v6.pm]: image
image

image

image
image

**Review by julsemaan — CHANGES_REQUESTED** [review]

File: lib/pf/dhcp/processor_v6.pm (hunk 9 above)
Comment (dwlfrth) [reply lib/pf/dhcp/processor_v6.pm]: Holly crap !

File: lib/pf/dhcp/processor.pm (hunk 7 above)
Comment (dwlfrth) [reply lib/pf/dhcp/processor.pm:151]: See https://github.com/inverse-inc/packetfence/pull/2086/files#diff-d25e2fd577726e1d2e0e6a1ea4b499fbR264

**Comment by dwlfrth** [comment]
Fixed, review me

**Pull request merged**